src/metabase/lib/js.cljs

```diff
@@ -339,6 +339,41 @@
                                     ;; match up. Therefore de-dupe with `frequencies` rather than simply `set`.
                                     (assoc inner-query :fields (frequencies fields)))))))
 
+(defn- compare-legacy-field-refs
+  [[key1 id1 opts1]
+   [key2 id2 opts2]]
+  ;; A mismatch of `:base-type` or `:effective-type` when both x and y have values for it is a failure.
+  ;; If either ref does not have the `:base-type` or `:effective-type` set, that key is ignored.
+  (letfn [(clean-opts [o1 o2]
+            (not-empty
+              (cond-> o1
+                (not (:base-type o2))      (dissoc :base-type)
+                (not (:effective-type o2)) (dissoc :effective-type))))]
+    (= [key1 id1 (clean-opts opts1 opts2)]
+       [key2 id2 (clean-opts opts2 opts1)])))
+
+(defn- query=* [x y]
+  (cond
+    (and (vector? x)
+         (vector? y)
+         (= (first x) (first y) :field))
```

Comment:
What about checking that the options element is missing or a map? `compare-legacy-field-refs` blows up if it tries to `dissoc` from a number or string, for example.

Reply:
Garbage in, garbage out. Sending a query _that_ badly malformed may as well throw; returning true or false is worse at that point. It works correctly on maps and `nil`/missing.

Reply:
Fair enough, if you know those points. (If I see such a generic function I prefer not to blow up, as I usually don't know if the part that's causing the problem is essential at all.)

Reply:
Do you feel the same way now that it's a much more targeted `valid-filter-for?` rather than `compatible-type?` or `assignable?`?

Reply:
Sorry, I'm lost. How is this related to `valid-filter-for?`?